src/sparsezoo/v2/file.py

```diff
@@ -0,0 +1,217 @@
+
+import json
+import logging
+import os
+from typing import Any, Dict, Optional
+
+import onnx
+import yaml
+
+from PIL import Image
+from sparsezoo.utils.downloader import download_file
+from sparsezoo.utils.numpy import load_numpy_list
+
+
+__all__ = ["File"]
+
+
+class File:
+    """
+    Object to wrap around common files. Currently, supporting:
+    - numpy files
+    - onnx files
+    - markdown files
+    - json files
+    - csv files
+    - image files
+
+    :param name: name of the File
+    :param path: path of the File
+    :param url: url of the File
+    """
+
+    def __init__(
+        self, name: str, path: Optional[str] = None, url: Optional[str] = None
+    ):
+        self.name = name
+        self.url = url
+        self.path = path
+
+        # self.path can have any extension, including no extension.
+        # However, the File object also contains information
+        # About its loadable extensions.
+        # Loadable files can be read into the memory.
+
+        self.loadable_extensions = {
+            ".npz": self._validate_numpy,
+            ".onnx": self._validate_onnx,
+            ".md": self._validate_markdown,
+            ".json": self._validate_json,
+            ".csv": self._validate_csv,
+            ".jpg": self._validate_img,
+            ".png": self._validate_img,
+            ".jpeg": self._validate_img,
+            ".yaml": self._validate_yaml,
+        }
+
+    @classmethod
+    def from_dict(cls, file: Dict[str, Any]) -> "File":
+        """
+        Factory method for creating a File class object
+        from a file dictionary
+        (useful when working with the `request_json` from NeuralMagic).
+        :param file: a dictionary which contains an information about
+            the file (as returned by NeuralMagic API)
+        :return: File class object
+        """
+        name = file.get("display_name")
+        path = file.get("path")
+        url = file.get("url")
+
+        return cls(
+            name=name,
+            path=path,
+            url=url,
+        )
+
+    def download(self, destination_path: str, overwrite: bool = True):
+        """
+        Download the contents of the file from the url.
+
+        :param destination_path: the local file path to save the downloaded file to
+        :param overwrite: True to overwrite any previous files if they exist,
+            False to not overwrite and raise an error if a file exists
+        """
+        new_file_path = os.path.join(destination_path, self.name)
+
+        if self.url is None:
+            raise ValueError(
+                "The File object requires a valid attribute `url` to download "
+                "the file contents from. However, `url` is None."
+            )
+
+        if self.path is not None:
+            logging.warning(
+                f"Overwriting the current location of the File: {self.path} "
+                f"with the new location: {new_file_path}."
+            )
+
+        download_file(
```

Comment:
I have a suggestion,
`download_file` in it's current implementation doesn't seem robust enough, What if the download fails? or we get `403`, `504` kind of errors,

Would it make sense, to wrap either `download_file` or `download` function here in a `retry` decorator?

The idea being wherever network is involved we try a few times before throwing an error to the user. My idea is something like the following from [`tenacity`](https://tenacity.readthedocs.io/en/latest/), we could implement a smaller version of this decorator on our own:

 ```python
 @retry(stop=(stop_after_delay(10) | stop_after_attempt(5)))
def stop_after_10_s_or_5_retries():
    print("Stopping after 10 seconds or 5 retries")
    raise Exception
 ``` 
 
 Does this make sense here?